---
title: "RNotebook: In vitro and In vivo Endoderm projection"
#author: Stefano Davide Vianello
#date: 2021

output:
  html_notebook: 
  html_document:
    df_print: paged
    toc: true
    toc_float: true
  pdf_document: default
---

##SETUP PACKAGES AND LIBRARIES, CELL CYCLE ANNOTATIONS

Install the following packages onyour machine, if these have not been installed already
```{r}
# install required packages.
# Note: only need to run this cell if you have never run this code on this machine
install.packages('Seurat')
install.packages("devtools")
devtools::install_github(repo = 'mojaveazure/loomR', ref = 'develop')
install.packages('tidyverse')

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("AnnotationHub")
BiocManager::install("ensembldb")
BiocManager::install('multtest') 
install.packages('metap')

install.packages("stringr")
```

Load the libraries required to run this notebook

```{r, verbose=FALSE}
# Load libraries from installed packages
# Note: you need to run this cell every time you start a new R session
library(Seurat)
library(loomR)
library(dplyr)
library(ggplot2)

library(RCurl)
library(cowplot)

library(AnnotationHub)
library(ensembldb)

library(purrr)
library(RColorBrewer)

library(metap)

library(tibble)

library(stringr)
library(here)
library(forcats)

```

We will also preemptively prepare gene annotations

```{r}
# Download cell cycle genes for organism at https://github.com/hbc/tinyatlas/tree/master/cell_cycle. Read it in with:
cell_cycle_genes <- read.csv(file = "E:/Mus_musculus.csv", header = TRUE)

# Connect to AnnotationHub
ah <- AnnotationHub()

# Access the Ensembl database for organism
ahDb <- query(ah, 
              pattern = c("Mus musculus", "EnsDb"), 
              ignore.case = TRUE)

# Acquire the latest annotation files
id <- ahDb %>%
        mcols() %>%
        rownames() %>%
        tail(n = 1)

# Download the appropriate Ensembldb database
edb <- ah[[id]]

# Extract gene-level information from database
annotations <- genes(edb, 
                     return.type = "data.frame")

# Select annotations of interest
annotations <- annotations %>%
        dplyr::select(gene_id, gene_name, seq_name, gene_biotype, description)

# remove un-needed variables
rm(ah,ahDb, edb, id)

# Extract IDs for mitochondrial genes
mt <- annotations %>%
        dplyr::filter(seq_name == "MT") %>%
        dplyr::pull(gene_name)


# Get gene names for Ensembl IDs for each gene
cell_cycle_markers <- dplyr::left_join(cell_cycle_genes, annotations, by = c("geneID" = "gene_id"))

# Acquire the S phase genes
s_genes <- cell_cycle_markers %>%
        dplyr::filter(phase == "S") %>%
        pull("gene_name")
        
# Acquire the G2M phase genes        
g2m_genes <- cell_cycle_markers %>%
        dplyr::filter(phase == "G2/M") %>%
        pull("gene_name")
```

Finally, expand the memory size of your machine, to run the more computationally intensive parts of the code:
```{r}
#extend memory limit
memory.limit(size = 10000000000000)
```







## SUBSETTING OF GASTRULOID ENDODERM CLUSTERS

```{r}
#load processed Seurat object (Gastruloids)
#load("E://Rossi/Processing/D4D5D6D7/D4D5D6D7counts_final_integrated_clustered.RData") #10.5GB

# we are going to create a separate seurat object with just the cells of the clusters identified as the gut tube/endoderm
Gastruloid_endoderm <- subset(seurat_integrated, idents = c("13", "4"))

#save the dataset
#save(Gastruloid_endoderm, file="D://Rossi/RossiD4D5D6D7_endoderm.RData")
```

## SUBSETTING OF EMBYO ENDODERM CLUSTERS

```{r}
#load processed Seurat object (Gut tube)
#load("E://Nowotschin2019/E875counts_integrated_final_toexplore.RData") # this ia a 10.1GB file named: "sct_seurat"

# we are going to create a separate seurat object with just the cells of the clusters identified as the gut tube
Nowotschin_endoderm <- subset(sct_seurat, idents = c("2", "3", "6", "7", "8", "10","11","12", "13","14", "15", "20"))

#save the dataset
#save(Nowotschin_endoderm, file="D://Rossi/NowotschinE875_guttube.RData")
```


## LOAD AND COMBINE ENDODERM DATA

```{r}
# load the two datasets we are going to integrate: Gastruloid endoderm clusters, and E875 gut tube
load("E://Rossi/RossiD4D5D6D7_endoderm.RData") #loads object called "Gastruloid endoderm"
load("E://Rossi/NowotschinE875_guttube.RData") #loads object called "Nowotschin endoderm"

#extract the names of the cells belonging to each dataset
cells_gastruloid <- WhichCells(Gastruloid_endoderm)
cells_embryo <- WhichCells(Nowotschin_endoderm)

rm(Nowotschin_endoderm)
rm(Gastruloid_endoderm)

## from each complete dataset, only keep cells that belonged to the endoderm clusters
load("E://Rossi/Processing/D4D5D6D7/QCfiltered_D4D5D6D7_counts.RData") #loads object "filtered_seurat" of size 4GB
Gastruloid_endoderm <- subset(filtered_seurat, cells = cells_gastruloid)
rm(filtered_seurat)

load("E://Nowotschin2019/HPC_filtered.RData") #loads object "filtered_seurat" of size 11.6GB
Nowotschin_endoderm <- subset(filtered_seurat, cells = cells_embryo)
rm(filtered_seurat)

# save the onjects you just made, so you can load them directly later
# save(Gastruloid_endoderm, file="D://Rossi/Gastruloid_endoderm_clusters.RData")
# save(Nowotschin_endoderm, file="D://Rossi/Embryo_endoderm_clusters.RData")

rm(cells_gastruloid, cells_embryo)
```

We have now created raw seurat objects that only contain cells that would belong to "endoderm" clusters
```{r}
#load the endoderm cells from the two datasets:
#load("D://Rossi/Gastruloid_endoderm_clusters.RData")
#load("D://Rossi/Embryo_endoderm_clusters.RData")

#time to uniformise the metadata categories of both datasets so they can be merged later
metadata_g <- Gastruloid_endoderm@meta.data
#removes cells column, which might be giving some merging problems
metadata_g <- subset(metadata_g, select = -cells )
#uniformise column names
names(metadata_g)[names(metadata_g) == 'mitoRatio'] <- 'X_Mitochondrial_Content'
#get mitochondrial measures to same ratio
metadata_g$X_Mitochondrial_Content <- metadata_g$X_Mitochondrial_Content * 10
#update with the uniformised metadata
Gastruloid_endoderm@meta.data <- metadata_g
rm(metadata_g)

#MERGING THE DATASETS
# now we can merge the endoderm cells from the two datasets
merged_endoderm <- merge(x = Gastruloid_endoderm, 
                       y = Nowotschin_endoderm, 
                       add.cell.id = c("Gastruloid", "Embryo"))
rm(Gastruloid_endoderm, Nowotschin_endoderm)


head(merged_endoderm@meta.data)
tail(merged_endoderm@meta.data)
```

## Processing of the merged dataset (Gastruloid + endoderm)

```{r}
# Create metadata dataframe
metadata <- merged_endoderm@meta.data

# Add new cell IDs to metadata
metadata$cells <- rownames(metadata)

# Make new sample category (type of sample), by extracting the word from the cell name
metadata$Dataset <- NA
metadata$Dataset[which(str_detect(metadata$cells, "^Gastruloid_"))] <- "Gastruloid"
metadata$Dataset[which(str_detect(metadata$cells, "^Embryo_"))] <- "Embryo"
#and remove the "cells" column which we do not need anymore
metadata <- subset(metadata, select = -cells )

# Add metadata back to Seurat object
merged_endoderm@meta.data <- metadata
rm(metadata)

# Create .RData object to load at any time
#save(merged_endoderm, file = "E://Rossi/merged_endoderm.RData") #this is a 2.7GB file
```


```{r}
#load the processed Seurat object with both types of endoderm:
load("E://Rossi/merged_endoderm.RData") #loads object called "merged_endoderm", 2.7G.B

# set current sample identity to be that of the dataset (i.e. Gastruloid or Embryo)
Idents(object=merged_endoderm) <- "Dataset"

#Plot key QC parameters
VlnPlot(merged_endoderm, features = c("nCount_RNA", "nFeature_RNA", "X_Mitochondrial_Content"), ncol = 3,  pt.size = 0.25)
```

```{r}
#See number of cells per sample
merged_endoderm@meta.data %>% 
  	ggplot(aes(x=Dataset, fill=Dataset)) + 
  	geom_bar(stat = "count") +
    stat_count(geom = "text", colour = "white", size = 3.5,
    aes(label = ..count..),position=position_stack(vjust=0.5))+
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells")
```


```{r}
##Within our data we will have many genes with zero counts. These genes can dramatically reduce the average expression for a cell and so we will remove them from our data. First we will remove genes that have zero expression in all cells. Additionally, we will perform some filtering by prevalence. If a gene is only expressed in a handful of cells, it is not particularly meaningful as it still brings down the averages for all other cells it is not expressed in. For our data we choose to keep only genes which are expressed in 5 or more cells.

# Output a logical vector for every gene on whether the more than zero counts per cell
# Extract counts
counts <- GetAssayData(object = merged_endoderm, slot = "counts")

# Output a logical vector for every gene on whether the more than zero counts per cell
nonzero <- counts > 0

# Sums all TRUE values and returns TRUE if more than 5 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 5

# Only keeping those genes expressed in more than 5 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
merged_endoderm <- CreateSeuratObject(filtered_counts, meta.data = merged_endoderm@meta.data)

#clean up unused variables
rm(counts, filtered_counts, nonzero, keep_genes)
```


```{r}
##Final reassessment of QC metrics
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
merged_endoderm@meta.data %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=X_Mitochondrial_Content)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 10000) +
  	geom_hline(yintercept = 2000) +
  	facet_wrap(~Dataset)
```

```{r}
# Create .RData object to load at any time
#save(merged_endoderm, file="E://Rossi/merged_endoderm.RData") #overwrites previous instance
#load("E://Rossi/merged_endoderm.RData") #loads object "merged_endoderm" of size 2.7GB
```

## Normalization, variance stabilization, and regression of unwanted variation for each sample

```{r}
# NOTE: from the analysis of each single dataset I already know that I need to regress cell cycle score.
# I will not repeat it here for the combined dataset. This chunck would be the code required to check whether cells separate based on the phase of the cell cycle at which they are.

#Cell cycle scoring
################
#It is recommended to check the cell cycle phase before performing the sctransform method. Since the counts need to be comparable between cells and each cell has a different number of total UMIs, we do a rough normalization by dividing by total counts per cell and taking the natural log. This method isn't as accurate as the sctransform method that we will use ultimately to identify cell clusters, but it is sufficient to explore sources of variation in our data.

# Normalize the counts
seurat_phase <- NormalizeData(merged_endoderm)

# Perform cell cycle scoring (assumes you have extracted S and G2M features, see SETUP section)
seurat_phase <- CellCycleScoring(merged_endoderm,
                                   g2m.features = g2m_genes,
                                   s.features = s_genes)

# View cell cycle scores and phases assigned to cells  (now has cell ceycle scores)                               
View(seurat_phase@meta.data)      

#now we will lunch a very basic full analysis pipeline to see how phase differentiates cells
# Identify the most variable genes
seurat_phase <- FindVariableFeatures(seurat_phase, 
                     selection.method = "vst",
                     nfeatures = 2000, 
                     verbose = FALSE)

# Scale the counts
seurat_phase <- ScaleData(seurat_phase)

# Perform PCA
seurat_phase <- RunPCA(seurat_phase)

# Plot the PCA colored by cell cycle phase
DimPlot(seurat_phase,
        reduction = "pca",
        group.by= "Phase",
        split.by = "Phase"
        )
```


```{r}
#Normalisation, stabilisaiton, regression: SCTransform
################
#Now we can use the sctransform method as a more accurate method of normalizing, estimating the variance of the raw filtered data, and identifying the most variable genes. By default, sctransform accounts for cellular sequencing depth, or nUMIs. We already checked cell cycle and decided that it didn't represent a major source of variation in our data, but mitochondrial expression is another factor which can greatly influence clustering. Oftentimes, it is useful to regress out variation due to mitochondrial expression. However, if the differences in mitochondrial gene expression represent a biological phenomenon that may help to distinguish cell clusters, then we advise not regressing the mitochondrial expression.

# do these steps on the dataset
sct_seurat <- NormalizeData(merged_endoderm, verbose = TRUE)
sct_seurat <- CellCycleScoring(merged_endoderm, g2m.features=g2m_genes, s.features=s_genes)

# clean up unused variables
rm(merged_endoderm, g2m_genes, s_genes)

# run sctransform (will take some time)
sct_seurat <- SCTransform(sct_seurat, vars.to.regress = c("X_Mitochondrial_Content", "S.Score", "G2M.Score"), verbose = FALSE)

##Visualise dataset on UMAP
# These are now standard steps in the Seurat workflow for visualization and clustering
sct_seurat  <- RunPCA(sct_seurat , verbose = FALSE)
sct_seurat  <- RunUMAP(sct_seurat , dims = 1:30, verbose = FALSE)

sct_seurat  <- FindNeighbors(sct_seurat , dims = 1:30, verbose = FALSE)
sct_seurat  <- FindClusters(sct_seurat , verbose = FALSE)



# save the output as a ready to load RData object
#save(sct_seurat, file="E://Rossi/endoderm_sct_seurat.RData") #loads "sct_seurat" object of size 6GB

#plot
DimPlot(sct_seurat , label = TRUE, group.by = "Dataset")
```


```{r}
DimPlot(sct_seurat, group.by = "Dataset", cols = c("firebrick1", "gray35")) +labs(title = "UMAP (separated by batch)")

DimPlot(sct_seurat, group.by = "Dataset", cols = c("gray35", "firebrick1")) +labs(title = "UMAP (separated by batch)")

```

In this case, the two batches do not overlap at all, we need to integrate based on shared features


## INTEGRATION

```{r, fig.width=10, fig.height=6}
#If the cells cluster by sample, condition, dataset, or modality, this step can greatly improve your clustering and your downstream analyses. It can help to first run conditions individually if unsure what clusters to expect or expecting some different cell types between conditions (e.g. tumor and control samples), then run them together to see whether there are condition-specific clusters for cell types present in both conditions. Oftentimes, when clustering cells from multiple conditions there are condition-specific clusters and integration can help ensure the same cell types cluster together.

# Split seurat object by condition to perform cell cycle scoring and SCT on all samples
split_seurat <- SplitObject(sct_seurat, split.by = "Dataset")
split_seurat <- split_seurat[c("Embryo", "Gastruloid")]


# Create .RData object to load at any time
#save(split_seurat, file="D://Rossi/Endoderm_split_for_integration.RData")
#load("D://Rossi/Endoderm_split_for_integration.RData") #loads "split_seurat" of size 6GB

for (i in 1:length(split_seurat)) {
    split_seurat[[i]] <- NormalizeData(split_seurat[[i]], verbose = TRUE)
    split_seurat[[i]] <- CellCycleScoring(split_seurat[[i]], g2m.features=g2m_genes, s.features=s_genes)
    split_seurat[[i]] <- SCTransform(split_seurat[[i]], vars.to.regress = c("X_Mitochondrial_Content", "S.Score", "G2M.Score"), verbose = FALSE)
}

# Check which assays are stored in objects
#split_seurat$D4@assays

#Now we can see that in addition to the raw RNA counts, we now have a SCT component in our assays slot. The most variable features will be the only genes stored inside the SCT assay. As we move through the scRNA-seq analysis, we will choose the most appropriate assay to use for the different steps in the analysis.

#Generally, we always look at our cells before deciding whether we need to perform integration. If we had performed the normalization on both conditions together in a Seurat object and visualized the similarity between cells, we would have seen condition-specific clustering (see steps above)

###

#To integrate, we will use the shared highly variable genes from each condition identified using SCTransform, then, we will "integrate" or "harmonize" the conditions to overlay cells that are similar or have a "common set of biological features" between groups.

#Now, using our SCTransform object as input, let's perform the integration across conditions.

#First, we need to specify that we want to use all of the 3000 most variable genes identified by SCTransform for the integration. By default, this function only selects the top 2000 genes.

# Select the most variable features to use for integration
integ_features <- SelectIntegrationFeatures(object.list = split_seurat, nfeatures = 3000) 

# Prepare the SCT list object for integration
integrated_seurat <- PrepSCTIntegration(object.list = split_seurat, anchor.features = integ_features)

# Find best buddies - can take a while to run (will say 0% even when running)
integ_anchors <- FindIntegrationAnchors(object.list = integrated_seurat, 
                                        normalization.method = "SCT", 
                                        anchor.features = integ_features,
                                        k.filter = 200) # this value cannot be higher than the lower cell number in either sample

# Integrate across conditions
seurat_integrated <- IntegrateData(anchorset = integ_anchors, 
                                   normalization.method = "SCT")

## We will now process this new, integrated, object as usual

# Run PCA
seurat_integrated <- RunPCA(object = seurat_integrated)

# Run UMAP
seurat_integrated <- RunUMAP(seurat_integrated, 
                             dims = 1:40,reduction = "pca")


#all the steps above were done on the HPC cluster. Load output file:
#load("E://Rossi/endoderms_integrated.RData") #loads "seurat_integrated" file, of size 7.4GB

```


```{r}
# Plot PCA
PCAPlot(seurat_integrated,
        split.by = "Dataset", group.by= "Dataset") 


# Color by sample (batch) to check that they now do not separate anymore
DimPlot(seurat_integrated, group.by = "Dataset", cols = c("firebrick1", "gray35")) +labs(title = "Integrated UMAP (separated by origin)")&DarkTheme()

DimPlot(seurat_integrated, group.by = "Dataset", cols = c("gray35", "goldenrod1")) +labs(title = "Mapping of Gastruloid Endoderm")&DarkTheme()
```



```{r}
##Segregation of clusters by various sources of uninteresting variation

# Determine metrics to plot present in seurat_integrated@meta.data
#metrics <-  c("nCount_RNA", "nFeature_RNA", "S.Score", "G2M.Score", "X_Mitochondrial_Content")
metrics <-  c("nCount_RNA", "nFeature_RNA")
metrics <-  c("S.Score", "G2M.Score")
metrics <-  c("X_Mitochondrial_Content")

# and to visualise them run UMAP
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
#could you inprove by going back at the regression step and choose additional features to regress out?
```

```{r, fig.width = 14,  fig.height = 10}
#Exploration of the PCs driving the different clusters
#We can also explore how well our clusters separate by the different PCs; we hope that the defined PCs separate the cell types well. To visualize this information, we need to extract the UMAP coordinate information for the cells along with their corresponding scores for each of the PCs to view by UMAP.

#First, we identify the information we would like to extract from the Seurat object, then, we can use the FetchData() function to extract it.

# Defining the information in the seurat object of interest
columns <- c(paste0("PC_", 1:16),
            "ident",
            "UMAP_1", "UMAP_2")

# Extracting this data from the seurat object
pc_data <- FetchData(seurat_integrated, 
                     vars = columns)

# Adding cluster label to center of cluster on UMAP
umap_label <- FetchData(seurat_integrated, 
                        vars = c("ident", "UMAP_1", "UMAP_2"))  %>%
  group_by(ident) %>%
  summarise(x=mean(UMAP_1), y=mean(UMAP_2))
  
# Plotting a UMAP plot for each of the PCs
map(paste0("PC_", 1:16), function(pc){
        ggplot(pc_data, 
               aes(UMAP_1, UMAP_2)) +
                geom_point(aes_string(color=pc), 
                           alpha = 0.7) +
                scale_color_gradient(guide = FALSE, 
                                     low = "grey90", 
                                     high = "blue")  +
                geom_text(data=umap_label, 
                          aes(label=ident, x, y)) +
                ggtitle(pc)
}) %>% 
        plot_grid(plotlist = .)
```

```{r}
# Examine PCA results (here only showing first 16, but we used 40 for clustering)
print(seurat_integrated[["pca"]], dims = 1:16, nfeatures = 5)
```


## CLUSTERING

```{r}
#https://github.com/hbctraining/scRNA-seq/blob/master/lessons/07_SC_clustering_cells_SCT.md
DefaultAssay(seurat_integrated) <- "integrated"
# Explore heatmap of PCs
DimHeatmap(seurat_integrated, 
           dims = 1:9, 
           cells = 500, 
           balanced = TRUE)

# Printing out the most variable genes driving PCs
print(x = seurat_integrated[["pca"]], 
      dims = 1:10, 
      nfeatures = 5)

# Plot the elbow plot
ElbowPlot(object = seurat_integrated, 
          ndims = 60)
```

```{r}
##
# Determine the K-nearest neighbor graph
# only run if you want to use more of the 30 dims used before
seurat_integrated <- FindNeighbors(object = seurat_integrated, 
                                dims = 1:40)

# Determine the clusters for various resolutions                                
seurat_integrated <- FindClusters(object = seurat_integrated,
                               resolution = c(0.2, 0.4, 0.6, 0.8, 1.0))

# Assign identity of clusters (i.e. choose which resolution to consider)
Idents(object = seurat_integrated) <- "integrated_snn_res.0.4" #check the object metadata to know this label
```


## REASSIGNING CLUSTER IDENTITIES
We can now re-label Gastruloid cells as belonging to either of the two original endoderm clusters

```{r}
#load the Seurat object of subsetted gastruloid endoderm
load("E://Rossi/RossiD4D5D6D7_endoderm.RData")

#get name of cells belonging to each of the two endoderm clusters
cells_c13 <- WhichCells(Gastruloid_endoderm, idents ="13")
cells_c13 <- paste0("Gastruloid_", cells_c13)
cells_c04 <- WhichCells(Gastruloid_endoderm, idents = "4")
cells_c04 <- paste0("Gastruloid_", cells_c04)
rm(Gastruloid_endoderm)

#add a column with the clusters
metadata <- seurat_integrated@meta.data

metadata$original.clusters <- NA
metadata[cells_c04, "original.clusters"]  = "4"
metadata[cells_c13, "original.clusters"]  = "13"

#reintegrate metadata into seurat object
seurat_integrated@meta.data <- metadata
```



B- FIND MARKERS OF CLUSTERS OF INTEREST

```{r}
#Find marker of each of the gut tube clusters
Idents(seurat_integrated) <- "integrated_snn_res.0.8"

markers_r0.8_09 <- FindMarkers(seurat_integrated,
                          ident.1 = 9,
                          ident.2 = c(0,1,2,3,4,5,6,7,8,10,11,12,13,14,15, 16, 17,18,19,20,21))   

markers_r0.8_09 <- markers_r0.8_09 %>% 
                           rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))


## I am finding marker genes as those that have high pc.1 and low pc.2.
## to facilitate sorting i will create a "delta" column with this difference
markers_r0.8_09 <- markers_r0.8_09 %>% 
  mutate(delta = (pct.1 - pct.2)) %>%
   dplyr::arrange(-delta) 

#------------------

markers_r0.8_10 <- FindMarkers(seurat_integrated,
                          ident.1 = 10,
                          ident.2 = c(0,1,2,3,4,5,6,7,8,9,11,12,13,14,15, 16, 17,18,19,20,21))   

markers_r0.8_10 <- markers_r0.8_10 %>% 
                           rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))

#row.names(markers_r0.8) <- as.character(markers_r0.8$gene2)

## I am finding marker genes as those that have high pc.1 and low pc.2.
## to facilitate sorting i will create a "delta" column with this difference
markers_r0.8_10 <- markers_r0.8_10 %>% 
  mutate(delta = (pct.1 - pct.2)) %>%
   dplyr::arrange(-delta) 

#------------------

markers_r0.8_11 <- FindMarkers(seurat_integrated,
                          ident.1 = 11,
                          ident.2 = c(0,1,2,3,4,5,6,7,8,9, 10,12,13,14,15, 16, 17,18,19,20,21))   

markers_r0.8_11 <- markers_r0.8_11 %>% 
                           rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))

#row.names(markers_r0.8) <- as.character(markers_r0.8$gene2)

## I am finding marker genes as those that have high pc.1 and low pc.2.
## to facilitate sorting i will create a "delta" column with this difference
markers_r0.8_11 <- markers_r0.8_11 %>% 
  mutate(delta = (pct.1 - pct.2)) %>%
   dplyr::arrange(-delta) 

#------------------
markers_r0.8_02 <- FindMarkers(seurat_integrated,
                          ident.1 = 2,
                          ident.2 = c(0,1,3,4,5,6,7,8,9, 10,11,12,13,14,15, 16, 17,18,19,20,21))   

markers_r0.8_02 <- markers_r0.8_02 %>% 
                           rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))

#row.names(markers_r0.8) <- as.character(markers_r0.8$gene2)

## I am finding marker genes as those that have high pc.1 and low pc.2.
## to facilitate sorting i will create a "delta" column with this difference
markers_r0.8_02 <- markers_r0.8_02 %>% 
  mutate(delta = (pct.1 - pct.2)) %>%
   dplyr::arrange(-delta) 

#------------------

markers_r0.8_00 <- FindMarkers(seurat_integrated,
                          ident.1 = 0,
                          ident.2 = c(1,2,3,4,5,6,7,8,9, 10, 11, 12,13,14,15, 16, 17,18,19,20,21))   

markers_r0.8_00 <- markers_r0.8_00 %>% 
                           rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))

#row.names(markers_r0.8) <- as.character(markers_r0.8$gene2)

## I am finding marker genes as those that have high pc.1 and low pc.2.
## to facilitate sorting i will create a "delta" column with this difference
markers_r0.8_00 <- markers_r0.8_00 %>% 
  mutate(delta = (pct.1 - pct.2)) %>%
   dplyr::arrange(-delta) 

#------------------

markers_r0.8_04 <- FindMarkers(seurat_integrated,
                          ident.1 = 4,
                          ident.2 = c(0,1,2,3,5,6,7,8,9, 10,11,12,13,14,15, 16, 17,18,19,20,21))   

markers_r0.8_04 <- markers_r0.8_04 %>% 
                           rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))

#row.names(markers_r0.8) <- as.character(markers_r0.8$gene2)

## I am finding marker genes as those that have high pc.1 and low pc.2.
## to facilitate sorting i will create a "delta" column with this difference
markers_r0.8_04 <- markers_r0.8_04 %>% 
  mutate(delta = (pct.1 - pct.2)) %>%
   dplyr::arrange(-delta) 

#------------------

markers_r0.8_05 <- FindMarkers(seurat_integrated,
                          ident.1 = 5,
                          ident.2 = c(0,1,2,3,4,6,7,8,9, 10,11,12,13,14,15, 16, 17,18,19,20,21))   

markers_r0.8_05 <- markers_r0.8_05 %>% 
                           rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))

#row.names(markers_r0.8) <- as.character(markers_r0.8$gene2)

## I am finding marker genes as those that have high pc.1 and low pc.2.
## to facilitate sorting i will create a "delta" column with this difference
markers_r0.8_05 <- markers_r0.8_05 %>% 
  mutate(delta = (pct.1 - pct.2)) %>%
   dplyr::arrange(-delta) 


#---------------
save(markers_r0.8_00, file="E://Nowotschin2019/Endoderms_r08_00.RData")
save(markers_r0.8_02, file="E://Nowotschin2019/Endoderms_r08_02.RData")
save(markers_r0.8_04, file="E://Nowotschin2019/Endoderms_r08_04.RData")
save(markers_r0.8_05, file="E://Nowotschin2019/Endoderms_r08_05.RData")
save(markers_r0.8_09, file="E://Nowotschin2019/Endoderms_r08_09.RData")
save(markers_r0.8_10, file="E://Nowotschin2019/Endoderms_r08_10.RData")
save(markers_r0.8_11, file="E://Nowotschin2019/Endoderms_r08_11.RData")
```


```{r}
#save file for ease of access
#save(seurat_integrated, file= "D://Rossi/Processing/endoderms_for_exploration.RData")
load("E://Rossi/Processing/endoderms_for_exploration.RData")
```

## EXPLORE DATA

```{r}
## Plot UMAP in grey and Gastruloid clusters on top
DimPlot(seurat_integrated, 
        group.by = "original.clusters", 
        cols = c("deepskyblue", "indianred1"), 
        na.value="gray85",
        pt.size = 0.9) +labs(title = "Gastruloid endoderm and Embryonic gut tube")
```


```{r, fig.width=10, fig.height=6}

## Plot UMAP with colored clusters

Idents(seurat_integrated) <- "integrated_snn_res.0.8"

# Plot again your clusters on the UMAP, for reference
plot <- DimPlot(seurat_integrated,
        reduction = "umap",
        label = F,
        label.size = 6,
        ) +labs(title = "UMAP with resolution 0.4")
 
LabelClusters(plot = plot, id = "ident", color = "black", size = 4)
```




#### PREPRINT FIGURES
Below are self-contained chuncks of codes to reproduce the figures (or elements of figures) published on the preprint (Vianello and Lutolf, "In vitroendoderm emergence andself-organisation in the absence ofextraembryonic tissues and embryonicarchitecture")


```{r}
#save file for ease of access
#save(seurat_integrated, file= "D://Rossi/Processing/endoderms_for_exploration.RData")
load("E://Rossi/Processing/endoderms_for_exploration.RData")
```

#### GUT TUBE UMAP (FIG 8A, TOP)

```{r, fig.width=10, fig.height=6}
Idents(seurat_integrated) <- "integrated_snn_res.0.8"

colors = c("gray85", "gray65", "gray75", "gray85", "gray75", "gray55", "gray65",
           "gray55", "gray75", "gray55","gray55", "gray85", "gray75", "gray65",
           "gray65", "gray75","gray65", "gray75", "gray75","gray65", "gray75", 
           "gray75")

# Plot again your clusters on the UMAP, for reference
plot <- DimPlot(seurat_integrated,
        reduction = "umap",
        label = F,
        label.size = 6,
        pt.size=0.9,
        cols = colors,
        ) +labs(title = "UMAP with resolution 0.8")
 
LabelClusters(plot = plot, id = "ident", color = "black", size = 2)
```

#### FEATUREPLOTS OF ANTERIOR AND POSTERIOR GUT TUBE MARKERS (FIG 8A, BOTTOM)

```{r, fig.width=10, fig.height=6}
plotted_gene = c("T", "Cdx2", "Sox2", "Nkx2-1", "Nkx2-5","Pax9","Pyy", "Pdx1")

plotted_color =  colorRampPalette(brewer.pal(10,"Spectral"))(length(plotted_gene))

 
for (i in seq(length(plotted_gene))) {
 plot <- FeaturePlot(seurat_integrated, 
                     reduction = "umap", 
                     features = plotted_gene[i], 
                     cols = c("gray85", "black"), 
                     max.cutoff = 1,
                     pt.size = 1,
                     order = TRUE)+
    labs(title = plotted_gene[i])+ 
    theme(plot.title = element_text(size=40))
 print(plot)
}
```

#### GUT TUBE WITH GASTRULOID CELLS IN RED AND BLUE (FIG 8B, TOP)

```{r}
DimPlot(seurat_integrated, 
        group.by = "original.clusters", 
        cols = c("deepskyblue", "indianred1"), 
        na.value="gray85",
        pt.size = 0.9) +labs(title = "Gastruloid endoderm and Embryonic gut tube")
```

#### BARPLOT WITH DISTRIBUTION OF GASTRULOID CELLS ALONG GUT TUBE CLUSTERS (FIG 8B, BOTTOM)

```{r}
#make a dataframe with the data required
df <- table(seurat_integrated@meta.data$integrated_snn_res.0.8, seurat_integrated@meta.data$original.clusters)
df <- as.data.frame.matrix(df) 
df <- df %>% 
  rownames_to_column(var = "cluster")

#convert to tidydata for plotting
new_df <- data.frame("cluster" = df$cluster,
                 "cell_number" = c(df$"13", df$"4") ,
                 "endoderm_cluster" = rep(c("13", "4"), each=22))

#plot the data
p <- ggplot(data=new_df, aes(x=cluster, y=cell_number, fill=endoderm_cluster)) +
  geom_bar(stat="identity",color="black", width=0.7, position=position_dodge())+ ggtitle("Composition of the FoxA2+ population across samples") 

p+scale_fill_manual(values=c("magenta", "green"))&DarkTheme()+ theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))


#reorder gut tube clusters so they match AP order
p <- new_df %>%
  mutate(cluster = fct_relevel(cluster, 
                            "9", "11", "18","13","10", "14", "8", "12", "20", "19", "3", "4", "1", "15", "5", "0", "17", "16", "2", "7", "21", "6")) %>%
  ggplot( aes(x=cluster, y=cell_number, fill=endoderm_cluster)) +
    geom_bar(stat="identity",color="black", width=0.7, position=position_dodge())+
    ggtitle("Distribution of Gastruloid cells along the gut tube")+
    xlab("Gut tube clusters")

p+scale_fill_manual(values=c("deepskyblue", "indianred1"))+ theme_classic()

```


#### HIGHLIGHT OF SPECIFIC CLUSTERS WITHIN THE GUT TUBE UMAP (FIG8C)

Clusters with higer "mature endoderm"
```{r}
Idents(object=seurat_integrated) <- "integrated_snn_res.0.8"

colors = c("gray85", "gray85", "mediumpurple1", "gray85", "gray85", "gray85", "gray85",
           "gray85", "gray85", "plum2","lightgreen", "tan1", "gray85", "gray85",
           "gray85", "gray85","gray85", "gray85", "gray85","gray85", "gray85", 
           "gray85")

# Plot again your clusters on the UMAP, for reference
plot <- DimPlot(seurat_integrated,
        reduction = "umap",
        label = F,
        cols = colors,
        label.size = 6,
        pt.size = 1,
        ) +labs(title = "UMAP with resolution 0.8")
plot
```

Clusters with equal contributions
```{r}
Idents(object=seurat_integrated) <- "integrated_snn_res.0.8"

colors = c("goldenrod1", "gray85", "gray85", "gray85", "gray85", "gray85", "gray85",
           "gray85", "gray85", "gray85","gray85", "gray85", "gray85", "gray85",
           "gray85", "gray85","gray85", "gray85", "gray85","gray85", "gray85", 
           "gray85")

# Plot again your clusters on the UMAP, for reference
plot <- DimPlot(seurat_integrated,
        reduction = "umap",
        label = F,
        cols = colors,
        label.size = 6,
        pt.size = 1,
        ) +labs(title = "UMAP with resolution 0.8")
plot
```

Clusters with higer "early endoderm"
```{r}
Idents(object=seurat_integrated) <- "integrated_snn_res.0.8"

colors = c("gray85", "gray85", "gray85", "gray85", "lightgreen", "rosybrown1", "gray85",
           "gray85", "gray85", "gray85","gray85", "gray85", "gray85", "gray85",
           "gray85", "gray85","gray85", "gray85", "gray85","gray85", "gray85", 
           "gray85")

# Plot again your clusters on the UMAP, for reference
plot <- DimPlot(seurat_integrated,
        reduction = "umap",
        label = F,
        cols = colors,
        label.size = 6,
        pt.size = 1,
        ) +labs(title = "UMAP with resolution 0.8")
plot
```

#### FEATUREPLOT OF PAX9 EXPRESSION IN THE GUT TUBE (FIG 8D)

```{r, fig.width=8, fig.height=6}

DefaultAssay(seurat_integrated) <- "RNA"
seurat_integrated <- NormalizeData(seurat_integrated, verbose = FALSE)

plot <- FeaturePlot(seurat_integrated, 
                     reduction = "umap", 
                     features = "Pax9", 
                     cols = c("gray85", "black"), 
                     max.cutoff = 1,
                     pt.size = 1,
                     order = TRUE)+
    labs(title = "Pax9")+ 
    theme(plot.title = element_text(size=40))

print(plot)

```

##### FeaturePlots of AP endoderm markers (all) (FigS1)

```{r, fig.width=8, fig.height=6}

DefaultAssay(seurat_integrated) <- "RNA"
seurat_integrated <- NormalizeData(seurat_integrated, verbose = FALSE)

# plot nowotschin AP genes
plotted_gene = c("Nkx2-1", "Pax8", "Nkx2-5", "Nkx2-3","Isl1", "Otx2", "Prrx2", "Six1", "Foxg1", "Irx3", "Hoxb1", "Meis2", "Gata6", "Foxa3", "Cdx2", "Hoxa7", "Hoxb8", "Hoxc8", "Hoxc9", "Tlx2")


plotted_color =  colorRampPalette(brewer.pal(20,"Spectral"))(length(plotted_gene))

for (i in seq(length(plotted_gene))) {
 plot <- FeaturePlot(seurat_integrated, 
                     reduction = "umap", 
                     features = plotted_gene[i], 
                     cols = c("gray85",plotted_color[i]),#cols = c("gray85", "black"), 
                     max.cutoff = 1,
                     pt.size = 1,
                     order = TRUE)+
    labs(title = plotted_gene[i])+ 
    theme(plot.title = element_text(size=40))
 print(plot)
}
```


```{r}
sessionInfo()
```

_RNotebook by Stefano Vianello (EPFL)_

stefano.vianello@epfl.ch 

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

